### ISSUE TITLE
Configure mutagen mode volumes using options instead of file sharing

### ISSUE THREAD

**Issue by kinghuang** [docker/for-mac#4590]
- [X] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior

Be able to use Mutagen-based caching using a volume mode like the existing `consistent`, `cached`, and `delegated` options.

### Actual behavior

Directories have to be manually added to File Sharing preferences, and then caching enabled.

### Information

Docker for Mac (Edge) 2.3.1.0 introduces a new Mutagen-based caching mechanism. Configuring volumes to use this is too complicated. Specific directories have to be added to the File Sharing preferences, and then caching manually turned on.

It would be better if a new option can be added alongside the existing caching options like `delegated`.

### Diagnostic logs

N/A

### Steps to reproduce the behavior

Follow the instructions on [Mutagen-based caching](https://docs.docker.com/docker-for-mac/mutagen-caching/).

**Comment by djs55** [comment]
Thanks for the report. I'll add investigating this to the backlog and include your suggestion of using `:delegated`.

**Comment by Fire-** [comment]
I'd also like to see this handled via volume configuration - it really doesn't make sense to need to tell other developers to perform specific client-configuration changes when we could otherwise have it happen automatically via our existing docker-compose.yml . 

Ideally this would be able to be a transparent performance improvement if configured correctly

**Comment by kinghuang** [comment]
There's quite a bit of discussions on #1592 about this.

**Comment by ajardin** [comment]
> Thanks for the report. I'll add investigating this to the backlog and include your suggestion of using `:delegated`.

Have you had a chance to test that possibility?

**Comment by djs55** [comment]
@ajardin yes, in fact I have an active PR to add `:delegated`. I'll share the build when it's in slightly better shape.

**Comment by ajardin** [comment]
> @ajardin yes, in fact I have an active PR to add `:delegated`. I'll share the build when it's in slightly better shape.

Awesome! Thanks again for your work. 😉

**Comment by djs55** [comment]
@ajardin the PR to enable caching on `:delegated` was merged and released in Edge 2.3.2.0: https://desktop.docker.com/mac/edge/46268/Docker.dmg -- let me know what you think.

The behaviour is now:
- `docker run -v /foo:/bar:delegated` will create the two-way sync the first time it's executed. A notification is displayed to warn the user it might take some time.
- subsequent `docker run -v /foo:/bar` will use this sync-- it's not necessary to keep adding `:delegated` but it's probably advisable

It's still an open question what the best behaviour of a `docker run -v /foo:/bar` is (i.e. without `:delegated`). Currently it will use the sync if it exists. Perhaps we should only use the sync if if `:delegated` is provided? Or possibly bypass the sync if `:consistent` is provided. Feedback welcome!

**Comment by daveisfera** [comment]
Is this also true of using `:delegated` in `docker-compose`?

**Comment by djs55** [comment]
Yes, `:delegated` in compose should set up the two-way sync too.

**Comment by ajardin** [comment]
> @ajardin the PR to enable caching on `:delegated` was merged and released in Edge 2.3.2.0: https://desktop.docker.com/mac/edge/46268/Docker.dmg -- let me know what you think.
> 
> The behaviour is now:
> 
> * `docker run -v /foo:/bar:delegated` will create the two-way sync the first time it's executed. A notification is displayed to warn the user it might take some time.
> * subsequent `docker run -v /foo:/bar` will use this sync-- it's not necessary to keep adding `:delegated` but it's probably advisable
> 
> It's still an open question what the best behaviour of a `docker run -v /foo:/bar` is (i.e. without `:delegated`). Currently it will use the sync if it exists. Perhaps we should only use the sync if if `:delegated` is provided? Or possibly bypass the sync if `:consistent` is provided. Feedback welcome!

I tested your build with my Docker Compose environment for Magento 2 projects, and it seems to work (again) like a charm! 🥳 

1. I checked that there are no caching instructions in Docker for Mac UI.
2. I added the `delegated` flag on mount-points for the project codebase.
3. When I started the Docker Compose environment, I saw a new entry in the Docker for Mac UI and the container creation waited for the end of the synchronization initialization. My files were visible in my container as soon as it was available.
5. I stopped and my Docker Compose environment and removed all existing containers.
6. When I started the Docker Compose environment, the container creation was almost instant. My files were visible in my container as soon as it was available.

From my point of view, we should only use the synchronization if the `delegated` flag is provided. It could be quite tricky to debug otherwise. But, how will you handle this case since you allow caching to be enabled from the Docker for the Mac UI?

**Comment by daveisfera** [comment]
> Yes, `:delegated` in compose should set up the two-way sync too.

Is there a way to verify that? (i.e. check that a mount is using mutagen?)

**Comment by djs55** [comment]
@daveisfera yes, if you `docker run -v ~/workspace:/workspace:delegated alpine sh` and then inspect the container with
```
$ docker inspect <name> | jq .[0].HostConfig.Binds
[
  "/var/lib/mutagen/desktop09ac2896f1b0722124996ec19378aab0:/workspace:delegated"
]
```
the source path(s) will be rewritten to `/var/lib/mutagen/...` if the two-way sync is enabled properly.

**Comment by daveisfera** [comment]
Then it looks like it's not working with `docker-compose`, because here what `docker inspect` says:
```
"HostConfig": {
    "Binds": [
        "/Users/dlj/.aws:/root/.aws:z,delegated",
        "/Users/dlj/projects/mycode:/usr/src/app:z,delegated"
    ],
```

**Comment by ajardin** [comment]
On my side, the result is quite different (still from my Docker Compose environment for Magento 2).

```
$ docker inspect deaf36de0db1 | jq -r ".[0].HostConfig.Binds"
[
  "/var/lib/mutagen/desktopfa91d43ad983097cfecaeaff6a384e83:/var/www/html:delegated",
  "/run/host-services/ssh-auth.sock:/run/host-services/ssh-auth.sock:rw"
]
```

**Comment by daveisfera** [comment]
I found the problem. I'm using `:z,delegated` so that the proper permissions are set for our Linux users and it appears that currently only `:delegated` enables this when multiple flags should be supported for use in different environments. For testing, I switched to `:delegated` to test this functionality, but then the container times out after about 2 minutes with a exit status of 137 (which I believe is OOM)

**Comment by andytson-inviqa** [comment]
I've created https://github.com/docker/for-mac/issues/4756 as there should be a choice in being able to use osxfs delegated volumes

**Comment by stephen-turner** [comment]
@daveisfera's issue has moved to #4718 (and is now fixed).

**Comment by djs55** [comment]
There's a new experimental build: https://desktop-stage.docker.com/mac/edge/46638/Docker.dmg which has a number of improvements including
- using `:consistent` will bypass all caching and make I/O on the volume fully synchronous
- using `:cached` will bypass any mutagen sync and behave the same as `:delegated` does in older builds. This is backwards compatible because in older builds `:cached` and `:delegated` have the same behaviour.

There are a few other improvements so I've listed them in a mini-changelog on this public gist: https://gist.github.com/djs55/0e8957e1e36431a43cbfd3caaffa9165

**Comment by daveisfera** [comment]
I agree with the concept of having it be controlled by options, but the current UI in Preferences is confusing (says it will restart but actually doesn't and such) and the real problem is that once a mount is created, it has to be removed through the UI. It should automatically go away when no longer used or at least have a way to easily clean up "orphaned" mounts so that extra caches can't end up hanging out all over the place.

**Comment by jasonwilliams** [comment]
Is it possible to override the docker-compose `:delegated` options for mutagen and set my own?
Since switching to the latest Edge i now see my whole project being synced (in the sharing files view), before i was only sharing directories i  manually added.

How do i revert or override this behaviour without changing our docker-compose.yml ?

**Comment by stephen-turner** [comment]
@jasonwilliams Can you explain your use case a bit more? It should be the same files that are being synced as before, although now using Mutagen.

**Comment by jasonwilliams** [comment]
@stephen-turner basically what happened here:
https://github.com/docker/for-mac/issues/4593#issuecomment-665745942

Since switching from an older build (https://desktop-stage.docker.com/mac/edge/45429/Docker.dmg) to the latest edge it now caches the whole directory, it didn't do this before. My docker-compose.yml uses `:delegated`. I've managed to circumvent this by using `:cached` and then setting the paths manually in the UI

**Comment by stephen-turner** [comment]
Could you share your docker compose file, and the differences that you're seeing (which specific directories have changed)? Thanks.

**Comment by jasonwilliams** [comment]
@stephen-turner I think its fine, according to https://github.com/docker/for-mac/issues/4590#issuecomment-658117926 this seems to be by design that `:cached` bypasses mutagen caching but :delegated doesn’t and that’s what I’m seeing, im guessing this changed recently because before :delegated didn’t trigger a mutagen cache.
